File: app/src/main/java/com/elitec/alejotaller/feature/product/domain/repository/ProductRepository.kt
package com.elitec.alejotaller.feature.product.domain.repository

import com.elitec.alejotaller.feature.product.domain.entity.Product
import kotlinx.coroutines.flow.Flow

interface ProductRepository {
    fun observeAll(): Flow<List<Product>>
    suspend fun getById(itemId: String): Product?
    suspend fun sync(): Result<Unit>

    /** Soft-hold: incrementa reserved en remoto + cache local. */
    suspend fun incrementReserved(productId: String, quantity: Int): Product?

    /** Rollback soft-hold: decrementa reserved (clamp >= 0). */
    suspend fun decrementReserved(productId: String, quantity: Int): Product?

    /**
     * Fuente de verdad: lee Appwrite por id y actualiza cache offline-first.
     * Usado por soft-hold concurrente y por stock:changed.
     */
    suspend fun refreshFromRemote(productId: String): Product?
}

File: web/src/core/feature/product/domain/repository/product.repository.ts
import type { Product } from "../entity/Product";

export interface ProductRepository {
    getAll(): Promise<Product[]>
    /** Cache local inmediato (opcional; offline-first). */
    getLocalAll?(): Promise<Product[]>
    getById(id: string): Promise<Product | null>
    /** Fuente de verdad remota para operaciones críticas de stock. */
    refreshFromRemote?(id: string): Promise<Product | null>
    getByCategory(categoryId: string): Promise<Product[]>
    create(product: Product): Promise<Product>
    update(id: string, product: Partial<Product>): Promise<Product>
    delete(id: string): Promise<void>

    /**
     * Core 1: incrementa reserved atómicamente en Appwrite y persiste
     * únicamente la respuesta remota en Dexie.
     */
    incrementReserved(id: string, quantity: number): Promise<Product | null>

    /** Core 1: decrementa reserved atómicamente en Appwrite (mínimo 0). */
    decrementReserved(id: string, quantity: number): Promise<Product | null>

    sync(): Promise<void>
}
